server/src/main/java/io/deephaven/server/partitionedtable/PartitionedTableServiceGrpcImpl.java

```diff
@@ -0,0 +1,126 @@
+        GrpcUtil.rpcWrapper(log, responseObserver, () -> {
+            final SessionState session = sessionService.getCurrentSession();
+
+            SessionState.ExportObject<Table> targetTable =
+                    ticketRouter.resolve(session, request.getTableId(), "tableId");
+
+            session.newExport(request.getResultId(), "resultId")
+                    .require(targetTable)
+                    .onError(responseObserver)
+                    .submit(() -> {
+                        PartitionedTable partitionedTable = targetTable.get().partitionBy(request.getDropKeys(),
+                                request.getKeyColumnNamesList().toArray(String[]::new));
+                        safelyExecute(() -> {
+                            responseObserver.onNext(PartitionByResponse.getDefaultInstance());
+                            responseObserver.onCompleted();
+                        });
+                        return partitionedTable;
+                    });
+
+        });
+    }
+
+    @Override
+    public void merge(MergeRequest request, StreamObserver<ExportedTableCreationResponse> responseObserver) {
+        GrpcUtil.rpcWrapper(log, responseObserver, () -> {
+            final SessionState session = sessionService.getCurrentSession();
+
+            SessionState.ExportObject<PartitionedTable> partitionedTable =
+                    ticketRouter.resolve(session, request.getPartitionedTable(), "partitionedTable");
+
+            session.newExport(request.getResultId(), "resultId")
+                    .require(partitionedTable)
+                    .onError(responseObserver)
+                    .submit(() -> {
+                        Table merged = partitionedTable.get().merge();
+                        safelyExecute(() -> {
+                            responseObserver
+                                    .onNext(ExportUtil.buildTableCreationResponse(request.getResultId(), merged));
+                            responseObserver.onCompleted();
+                        });
+                        return merged;
+                    });
+        });
+
+    }
+
+    @Override
+    public void getTable(GetTableRequest request, StreamObserver<ExportedTableCreationResponse> responseObserver) {
+        GrpcUtil.rpcWrapper(log, responseObserver, () -> {
+            final SessionState session = sessionService.getCurrentSession();
+
+            SessionState.ExportObject<PartitionedTable> partitionedTable =
+                    ticketRouter.resolve(session, request.getPartitionedTable(), "partitionedTable");
+            SessionState.ExportObject<Table> keys =
+                    ticketRouter.resolve(session, request.getKeyTableTicket(), "keyTableTicket");
+
+            session.newExport(request.getResultId(), "resultId")
+                    // Due to the multiple operations running and the explicit size check, explicit contents read, we
+                    // run under the lock
+                    .requiresSerialQueue()
```

Comment:
I sort of hate that we need to use `whereIn` for single get, thus requiring the serial queue and the UGP lock. Can we differentiate this a little bit? If the key table is static and only has one row, we could definitely use `constituentFor` which is concurrent-safe.

Reply:
will drop this line - if the key table is static, we can validate if it is one row, then read the values out into Object[] and use that to read the constituent table. If not, we need the ugp and do the whereIn, as below